File: testdata/taint_clean.go
package testdata

import (
	"database/sql"
	"net/http"
)

func handleSafeSearch(w http.ResponseWriter, r *http.Request) {
	id := r.URL.Query().Get("id")
	db, _ := sql.Open("postgres", "")
	// Parameterized query — not vulnerable.
	db.Exec("SELECT * FROM users WHERE id=$1", id)
}

File: testdata/taint_clean.py
from flask import request
import sqlite3

def safe_search():
    q = request.args.get("q")
    conn = sqlite3.connect("app.db")
    cursor = conn.cursor()
    # Parameterized query — not vulnerable.
    cursor.execute("SELECT * FROM items WHERE name=%s", (q,))
    return cursor.fetchall()
